### PR TITLE
HDDS-11607. Improve compatibility of listStatusLight

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Background:

- HDDS-7715 in 1.4.0 introduced `listStatusLight` request, used when listing keys in FSO buckets.  Latest OM version was 4 at that time.  OM version is not checked by client.
- HDDS-11414 in 2.0.0 introduced new OM version 8, and added client-side check to send `listStatusLight` only if OM version is >= 8, otherwise use older `listStatus` request.
- OM version = 8 cannot be backported to 1.4.x without backporting features that introduced versions 5-7.  Therefore OM 1.4.x is stuck at version 4.
- OM version = 4 may or may not support `listStatusLight`, depending on whether it is 1.4.0 or some pre-release snapshot version.
- List keys fails with 1.4.0 client and 1.3.0 or earlier OM.
- Backport of HDDS-11414 to 1.4.x fixed this compatibility bug by completely disabling `listStatusLight` in client, regardless of whether OM supports it or not.

This PR changes 2.0.0+ client to use `listStatusLight` with OM versions that support it (1.4.0 and newer).  It can also be backported to 1.4.x, since it does not introduce a new OM version.

https://issues.apache.org/jira/browse/HDDS-11607

## How was this patch tested?

Tested with the commit preceding HDDS-7715. Implement a lightweight listStatus API:

- Built docker image with 68393452332a2d881c34eb2c21feee4b1a8e58e9.
- Set log level of client to DEBUG.
- Successfully ran `xcompat` with the custom docker image as old version.
- Verified FSO read compatibility test passes.
- Verified client log to confirm fallback to `listStatus`:

```
[main] DEBUG rpc.RpcClient: Failed listStatusLight, reverting to listStatus
```

Compatibility with other cluster versions are tested by CI:
https://github.com/adoroszlai/ozone/actions/runs/11550948011

### CODE REVIEW

**Comment by adoroszlai** [comment]
@errose28 @sumitagrawl @tanvipenumudy @xichen01 can you please review?

**Review by errose28 — COMMENTED** [review]
Thanks for working on this @adoroszlai. Indeed the situation is quite confusing 😄 Can we keep master the way it is with the regular check for v8, and do a custom fix using v4 like this only on the 1.4.x branch? Anyways I'm not sure full client/server cross compat is expected between the 1.x and 2.x lines.

My concern with putting this change into master is that it will create confusion around this feature forever, when it really only occurs in a few scenarios. I think having a 2.x client fall back to the original list status implementation even when talking to a 1.4.x server is actually a better tradeoff than baking this workaround code into the master branch.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 1]
```diff
@@ -2316,16 +2318,28 @@ public List<OzoneFileStatusLight> listStatusLight(String volumeName,
...
-          numEntries, allowPartialPrefixes);
-    } else {
-      return ozoneManagerClient.listStatus(keyArgs, recursive, startKey,
-              numEntries, allowPartialPrefixes)
-          .stream()
-          .map(OzoneFileStatusLight::fromOzoneFileStatus)
-          .collect(Collectors.toList());
+    if (omVersion.compareTo(OzoneManagerVersion.MAYBE_LIGHTWEIGHT_LIST_STATUS) >= 0) {
+      try {
+        return ozoneManagerClient.listStatusLight(keyArgs, recursive, startKey,
```
Comment (xichen01) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:2323]: Maybe we can add caching of whether OM supports `listStatusLight`, the current implementation would at worst probably make every one `listStatusLight` at least twice RPC with the OM.

Comment (adoroszlai) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:2323]: Good idea.  Please note that both version 4 (ad176eb0c10ced990d2063cb1b6d4004c77dbe55) and `listStatusLight` (6faf0023b057dc2cc51bdbf288a43e756fb42b4b) were both introduced in 1.4.0.  Fallback is necessary only for those using Ozone based on any of the 118 commits in between these two.

**Comment by xichen01** [comment]
LGTM +1

**Review by xichen01 — APPROVED** [review]

**Comment by sumitagrawl** [comment]
> Sure, I can abandon this change. We don't even need to backport it to 1.4.x, since `listStatusLight` is already disabled there ([6425679](https://github.com/apache/ozone/commit/6425679d71d414292153258841d3da79688b4ef2)). I think having V8 check in master and V4 check in 1.4.x is more confusing.
> 
> The only "ugly" part of this change is having to keep V8 reserved, but that is local to the `OzoneManagerVersion` class. Other than that, this is the way [HDDS-11414](https://issues.apache.org/jira/browse/HDDS-11414) should have fixed the situation.

Yes, agree, do not need this in master. But can have in 1.4.x version.

**Pull request closed without merging**